heron/api/src/java/com/twitter/heron/api/windowing/TriggerPolicy.java

```diff
@@ -80,4 +80,32 @@
    * @param state the state
    */
   void restoreState(S state);
+
+  /**
+   * Set the eviction policy to whatever eviction policy to use this with
+   *
+   * @param evictionPolicy the eviction policy
+   */
+  void setEvictionPolicy(EvictionPolicy<T, ?> evictionPolicy);
+
+  /**
+   * Set the trigger handler for this trigger policy to trigger
+   *
+   * @param triggerHandler the trigger handler
+   */
+  void setTriggerHandler(TriggerHandler triggerHandler);
+
+  /**
+   * Sets the window manager that uses this trigger policy
+   *
+   * @param windowManager the window manager
+   */
+  void setWindowManager(WindowManager<T> windowManager);
```

Comment:
I think we should remove this from the interface.  Having this interface is confusing to users since WindowManager implements the TriggerHandler interface.  WindowManager is also not a interface that users can implement thus it shouldn't be part of an interface that users will have to implement. I think having "setTriggerHandler" should suffice.  For build in triggers (e.g. WatermarkCountTriggerPolicy, WatermarkTimeTriggerPolicy), we can just pass in the WindowManager via their constructor.

Reply:
The reasoning behind this was that the WindowManager, TriggerHandler (which is the WindowManager) and TopologyConfig are three things the users may not have access to directly.  I was aiming to provide access to the same structures used in the internal TriggerPolicies to custom TriggerPolicies, should users wish to use them.  The idea was that, if someone wanted to, it should be possible to use the existing trigger policies in this manner.

Reply:
Looking more closely at the code I am confused how the a custom trigger policy will work.  All the existing trigger policies require the WindowManager to actually trigger windows. How will a user implementing a custom TriggerPolicy do this.  They have no reference to the WindowManager.  Shouldn't the windowManager be passed into the custom trigger implicitly?  As of right now, trigger polices can't trigger windows without the WindowManager.  That is why I suggested removing the setWindowManager interface, since trigger policies need the window manager regardless of what kind of trigger policy it is.

Reply:
Take a look at the changes i’ve made to windowedBoltExecutor.  Essentially, in order to allow these to attach to arbitrary triggerPolicies, Ive replaced their inclusion in the constructor with the setters you see in the interface for all trigger policies.

Reply:
I see